cdap-security/src/main/java/co/cask/cdap/security/store/FileSecureStore.java

```diff
@@ -0,0 +1,463 @@
+   */
+  @Override
+  public SecureStoreData get(String name) throws IOException {
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException(name + " not found in the secure store.");
+      }
+      Key key = keyStore.getKey(name, password);
+      return ((KeyStoreEntry) key).getData();
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the key " + name, e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * Returns the metadata for the element identified by the given name.
+   * @param name Name of the element
+   * @return An object representing the metadata associated with the element
+   */
+  private SecureStoreMetadata getSecureStoreMetadata(String name) throws IOException {
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException("Metadata for " + name + " not found in the secure store.");
+      }
+      Key key = keyStore.getKey(name, password);
+      return ((KeyStoreEntry) key).metadata;
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the metadata for " + name, e);
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  private static Path constructNewPath(Path path) {
+    return path.resolveSibling(path + "_NEW");
+  }
+
+  private static void loadFromPath(KeyStore keyStore, Path path, char[] password)
+    throws IOException {
+    try (InputStream in = new DataInputStream(Files.newInputStream(path))) {
+      keyStore.load(in, password);
+    } catch (NoSuchAlgorithmException | CertificateException e) {
+      throw new IOException("Unable to load the Secure Store. ", e);
+    }
+  }
+
+  private static boolean isBadOrWrongPassword(IOException ioe) {
+    /* According to Java keystore documentation if the load failed due to bad password
+       then the cause of the exception would be set to "UnrecoverableKeyException".
+       Unfortunately that is not the observed behavior. */
+    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+      return true;
+    }
+    // Workaround
+    return (ioe.getCause() == null)
+      && (ioe.getMessage() != null)
+      && ((ioe.getMessage().contains("Keystore was tampered")) || (ioe
+      .getMessage().contains("password was incorrect")));
+  }
+
+  /**
+   * Initialize the keyStore.
+   *
+   * @throws IOException If there is a problem reading or creating the keystore.
+   */
+  private static KeyStore locateKeystore(Path path, final char[] password) throws IOException {
+    Path newPath = constructNewPath(path);
+    KeyStore ks;
+    try {
+      ks = KeyStore.getInstance(SCHEME_NAME);
+      if (Files.exists(path)) {
+        // If the main file exists then the new path should not exist.
+        // Both existing means there is an inconsistency.
```

Comment:
No, _NEW means that we were able to persist the key store but the atomic rename failed.